engine/runtime/src/test/scala/org/enso/compiler/test/ModifiedTest.scala

```diff
@@ -0,0 +1,122 @@
+package org.enso.compiler.test
+
+import org.apache.commons.io.FileUtils
+import org.apache.commons.io.filefilter.{IOFileFilter, TrueFileFilter}
+import org.enso.interpreter.test.{InterpreterException, ValueEquality}
+import org.enso.pkg.PackageManager
+import org.enso.polyglot.{LanguageInfo, PolyglotContext, RuntimeOptions}
+import org.graalvm.polyglot.{Context, Value}
+import org.scalatest.BeforeAndAfterAll
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import java.io.{ByteArrayOutputStream, File, FileFilter}
+import java.nio.file.{Files, Paths, StandardCopyOption}
+
+trait ModifiedTest
+    extends AnyFlatSpec
+    with Matchers
+    with ValueEquality
+    with BeforeAndAfterAll {
+  val output                            = new ByteArrayOutputStream()
+  private[this] var testDirectory: File = _
+
+  def evalTestProject(name: String, iteration: Int): Value = {
+    val pkgPath =
+      new File(getClass.getClassLoader.getResource(name).getPath)
+
+    val testPkgPath = testDirectory.toPath.resolve(name).toFile
+    if (!testPkgPath.exists()) {
+      initialCopy(pkgPath, testPkgPath)
+    }
+    copyIterationFiles(pkgPath, testPkgPath, iteration)
+    val pkg        = PackageManager.Default.fromDirectory(testPkgPath).get
+    val mainFile   = pkg.mainFile
+    val mainModule = pkg.moduleNameForFile(mainFile)
+    val context = Context
+      .newBuilder(LanguageInfo.ID)
+      .allowExperimentalOptions(true)
+      .allowAllAccess(true)
+      .option(RuntimeOptions.PROJECT_ROOT, testPkgPath.getAbsolutePath)
+      .option(
+        RuntimeOptions.LANGUAGE_HOME_OVERRIDE,
+        Paths
+          .get("../../test/micro-distribution/component")
+          .toFile
+          .getAbsolutePath
+      )
+      .option(RuntimeOptions.EDITION_OVERRIDE, "0.0.0-dev")
+      .option(RuntimeOptions.STRICT_ERRORS, "true")
+      .option(RuntimeOptions.DISABLE_IR_CACHES, "false")
+      .out(output)
+      .in(System.in)
+      .option(RuntimeOptions.LOG_LEVEL, "WARNING")
+      .logHandler(System.err)
+      .build()
+    context.initialize(LanguageInfo.ID)
+    val executionContext = new PolyglotContext(context)
+    InterpreterException.rethrowPolyglot {
+      val topScope        = executionContext.getTopScope
+      val mainModuleScope = topScope.getModule(mainModule.toString)
+      val assocCons       = mainModuleScope.getAssociatedType
+      val mainFun         = mainModuleScope.getMethod(assocCons, "main").get
+      mainFun.execute()
+    }
+  }
+
+  private def initialCopy(from: File, to: File): Unit = {
+    val filter = new FileFilter {
+      override def accept(pathname: File): Boolean = {
+        !pathname.getName.matches(".+_(\\d+)\\.enso")
+      }
+    }
+    copySources(from, to, filter)
+  }
+
+  private def copyIterationFiles(from: File, to: File, iteration: Int): Unit = {
```

Comment:
This iteration logic is a bit complicated to me.

I understand that we match all files containing `_`? This seems possibly a bit too much. For now it works, but in case we wanted to re-use this class in the future - maybe we can add some special substring, like `Iterated` to the module name to mark it so? Or put it in a subdirectory `Iterations`/`Iterated` or something?

This is a nitpick, as it works good, but reading this code I needed to give it some think to see what is going on. Maybe at least add a comment to this function explaining in short what logic it is supposed to implement? (That it will select files containing `_` and select only the ith file from such sets.

Reply:
Not entirely. It matches all files ending with `_<digit>+.enso` suffix which should limit it. Besides, this only applies to projects specified in the test suite. I would prefer not to have a separate directory (at least not for this case) as it becomes harder to manage the contents of the project. 
I added some comments to clarify the logic but I'd prefer not to spend more time on it. If new use-cases arise it will be rather simple to change this.

Reply:
Sure, it's already really cool. The comments are good, thank you :)